core/payments/ondemand/on_demand_ledger.go

```diff
@@ -23,37 +24,67 @@ import (
-// are some practical concurrency limits based on the logic behind on-demand payments:
-// 1. Dispersals must be accounted for *in order*, since cumulative payment may only increase. Example:
-// - Assume a user creates two on-demand dispersals, `A` and `B`
-// - `B` will have a higher cumulative payment, since it was created second
-// - If the dispersals are accounted for out of order (`B` then `A`), then the accounting for `A` will fail,
-// since cumulative payment from `B`->`A` would decrease
-// 2. Debit reversion is only possible for a dispersal if no subsequent dispersals have been accounted for. Example:
-// - Assume a user creates two on-demand dispersals, `A` and `B`
-// - If `A` is accounted for, the debit for `A` can be reverted *only if* `B` has not yet been accounted for
-// - If this is not respected, the entity doing the accounting will diverge from the correct ledger state
-//
-// TODO(litt3): There are some improvements that may be made to the on-demand logic in the future, to permit a larger
-// number of concurrent on-demand dispersals.
+// This is a goroutine safe struct.
 type OnDemandLedger struct {
 	// total deposits available for the account in wei
 	totalDeposits *big.Int
 	// price per symbol in wei
 	pricePerSymbol *big.Int
 	// minimum number of symbols to bill
 	minNumSymbols uint64
-	// stores the cumulative payment for this ledger in wei
-	cumulativePaymentStore CumulativePaymentStore
+
+	// an optional store to back the cumulative payment for this account
+	//
+	// if non-nil, the new cumulative payment value will be stored here after each debit
+	cumulativePaymentStore *CumulativePaymentStore
+
+	// the latest cumulative payment for the account
+	cumulativePayment *big.Int
+	// used to synchronize computation and optional storing of the cumulativePayment
+	lock sync.Mutex
+}
+
+// Creates a new OnDemandLedger, backed by a CumulativePaymentStore
+//
+// The CumulativePaymentStore is used in this constructor to get the current cumulative payment value. After each
+// debit, the latest cumulative payment will be stored in the CumulativePaymentStore.
+func OnDemandLedgerFromStore(
+	ctx context.Context,
+	totalDeposits *big.Int,
+	pricePerSymbol *big.Int,
+	minNumSymbols uint64,
+	cumulativePaymentStore *CumulativePaymentStore,
+) (*OnDemandLedger, error) {
+	if cumulativePaymentStore == nil {
+		return nil, errors.New("cumulativePaymentStore cannot be nil")
+	}
+
+	cumulativePayment, err := cumulativePaymentStore.GetCumulativePayment(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("get cumulative payment from store: %w", err)
+	}
+
+	return newOnDemandLedger(totalDeposits, pricePerSymbol, minNumSymbols, cumulativePaymentStore, cumulativePayment)
+}
+
+// Creates a new OnDemandLedger, which *isn't* backed by a CumulativePayment store: the only representation of the
+// cumulative payment is in memory.
+func OnDemandLedgerFromValue(
```

Comment:
hmm is this really necessary? Perhaps call it `OnDemandLedgerMemstore"? FromValue sounds like its a normal constructor, not only used for tests.

Reply:
It is a normal constructor- this is what a client will use to instantiate this.

Reply:
20fe052934395b734d04edc5c84c950e6310260a

I added some additional detail in this commit. LMK what you think

Reply:
Oh wasn't thinking about the clients at all. Doc additions clear it up, thanks